numba/core/event.py

```diff
@@ -0,0 +1,407 @@
+    element is the event.
+    """
+    def __init__(self):
+        self.buffer = []
+
+    def on_start(self, event):
+        self.buffer.append((time.time(), event))
+
+    def on_end(self, event):
+        self.buffer.append((time.time(), event))
+
+
+@contextmanager
+def install_listener(kind, listener):
+    """Install a listener temporarily within the duration of the context.
+
+    Returns
+    -------
+    res : Listener
+        The *listener* provided.
+
+    Examples
+    --------
+
+    >>> with install_listener("numba:compile", listener):
+    >>>     some_code()  # listener will be active here.
+    >>> other_code()     # listener will be unregistered by this point.
+
+    """
+    register(kind, listener)
+    try:
+        yield listener
+    finally:
+        unregister(kind, listener)
+
+
+@contextmanager
+def install_timer(kind, callback):
+    """Install a TimingListener temporarily to measure the duration for
+    an event.
+
+    If the context completes successfully, the *callback* function is executed.
+    The *callback* function is expected to take a float argument for the
+    duration in seconds.
+
+    Returns
+    -------
+    res : TimingListener
+
+    Examples
+    --------
+
+    This is equivalent to:
+
+    >>> with install_listener(kind, TimingListener()) as res:
+    >>>    ...
+    """
+    tl = TimingListener()
+    with install_listener(kind, tl):
+        yield tl
+
+    if tl.done:
+        callback(tl.duration)
+
+
+@contextmanager
+def install_recorder(kind):
+    """Install a RecordingListener temporarily to record all events.
+
+    The buffer is filled from the context is closed.
+
+    Returns
+    -------
+    res : RecordingListener
+
+    Examples
+    --------
+
+    This is equivalent to:
+
+    >>> with install_listener(kind, RecordingListener()) as res:
+    >>>    ...
+    """
+    rl = RecordingListener()
+    with install_listener(kind, rl):
+        yield rl
+
+
+def start_event(kind, data=None):
+    """Signal the start of an event of *kind* with *data*.
+
+    Parameters
+    ----------
+    kind : str
+        Event kind.
+    data : any; optional
+        Extra event data.
+    """
+    evt = Event(kind=kind, status=EventStatus.START, data=data)
+    broadcast(evt)
+
+
+def end_event(kind, data=None, exc_details=None):
+    """Signal the end of an event of *kind*, *exc_details*.
+
+    Parameters
+    ----------
+    kind : str
+        Event kind.
+    data : any; optional
+        Extra event data.
+    """
+    evt = Event(
+        kind=kind, status=EventStatus.END, data=data, exc_details=exc_details,
+    )
+    broadcast(evt)
+
+
+@contextmanager
+def enter_event(kind, data=None):
+    """A context manager to signal the start and end events of *kind* with
+    *data*.
+
+    Parameters
+    ----------
+    kind : str
+        Event kind.
+    data : any; optional
+        Extra event data.
+    """
+    with ExitStack() as scope:
```

Comment:
Would `try/finally` suffice?

Reply:
was using `try/finally` until I needed extra code to fish out the exception. The `ExitStack` let's me push a function like a `__exit__()` to capture the 3-tuple exception infos.

Reply:
ah, I see, seems like this is a good choice then.